include/cuco/detail/static_map.inl

```diff
@@ -64,22 +72,27 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
 
   // TODO: memset an atomic variable is unsafe
   static_assert(sizeof(std::size_t) == sizeof(atomic_ctr_type));
-  CUCO_CUDA_TRY(cudaMemsetAsync(num_successes_, 0, sizeof(atomic_ctr_type)));
+  CUCO_CUDA_TRY(cudaMemsetAsync(num_successes_, 0, sizeof(atomic_ctr_type), stream));
   std::size_t h_num_successes;
 
   detail::insert<block_size, tile_size>
-    <<<grid_size, block_size>>>(first, first + num_keys, num_successes_, view, hash, key_equal);
+    <<<grid_size, block_size, 0, stream>>>(first, first + num_keys, num_successes_,
+                                                  view, hash, key_equal);
   CUCO_CUDA_TRY(cudaMemcpyAsync(
-    &h_num_successes, num_successes_, sizeof(atomic_ctr_type), cudaMemcpyDeviceToHost));
-  CUCO_CUDA_TRY(cudaDeviceSynchronize());
+    &h_num_successes, num_successes_, sizeof(atomic_ctr_type), cudaMemcpyDeviceToHost,
+    stream));
+  // stream'd execution assumes sync not required
+  if (stream == NULL)
+    CUCO_CUDA_TRY(cudaDeviceSynchronize()); // ensures legacy behavior
```

Comment:
```suggestion
    CUCO_CUDA_TRY(cudaStreamSynchronize(stream)); // ensures legacy behavior
```
We should replace all device sync with stream sync

Reply:
I am fine with replacing deviceSync with streamSync, however this change may reveal unintentional data race in applications that unknowingly benefitted from the device level synchronization provided by cudaDeviceSynchronize()

Reply:
I don't think we need to worry about people inadvertently relying on the device sync here. They should not have been doing that in the first place.